### DISCUSSION TITLE
Updated Examples

### DISCUSSION THREAD

**Discussion by bdkjones** [sparkle-project/Sparkle#2360] [General]
The existing examples in Sparkle's codebase are old. It would be great to have modern ones written in Swift that show the flow through `SPUUserDriver`--specifically the `async` variations of those protocol methods.

Working with the reply-block versions of those APIs is fairly straightforward, but it's less clear how to use the `async` versions, where we need to display a UI and await the user clicking a button to respond. The most common case, at least for the next decade, is an AppKit-based app that wraps a SwiftUI view in `NSHostingView` or `NSHostingViewController` to present that UI and it's not very obvious how to fit async/await into that flow.

Is there already an open issue for that enhancement?

**Comment by zorgiepoo** [comment]
> The existing examples in Sparkle's codebase are old. It would be great to have modern ones written in Swift that show the flow through SPUUserDriver--specifically the async variations of those protocol methods.

I'm assuming in the repository this refers to Sparkle's standard user interface (`SPUStandardUserDriver`), sparkle-cli (`SPUCommandLineUserDriver`), and the test app's alternative UI (`SUPopUpTitlebarUserDriver`). Note these applications are not primarily demonstration based (I bring this up because I'm not sure if we should have 'samples' in the main repo).

I think Swift's `async` / `await` support offers some syntax sugar to what is already possible. I haven't played around with it too much myself. Only a couple of the `SPUUserDriver` methods get exposed with an `async` variation I believe due to convention. All the `SPUUserDriver` methods are asynchronous friendly in nature as far as I can see (but not all of them are necessarily convertible to Swift's async/await).

There isn't too much documentation on implementing alternative user interfaces besides the API documentation for SPUUserDriver and other than looking for how the main repository does it in a few places. I suppose that could be improved. Most people still stick to the standard UI so it may be an "advanced" use case.
